malt/data/dataset.py

```diff
@@ -159,44 +164,91 @@ def featurize_all(self):
     def y(self):
         return [point.y for point in self.points]
 
-    @staticmethod
-    def batch_of_g_and_y(points):
-        # initialize results
-        gs = []
-        ys = []
+    @property
+    def smiles(self):
+        return [point.smiles for point in self.points]
 
-        # loop through the points
-        for point in points:
-            if not point.is_featurized():  # featurize
-                point.featurize()
-            if point.y is None:
-                raise RuntimeError("No data associated with data. ")
-            gs.append(point.g)
-            ys.append(point.y)
+    def _construct_extra(self):
+        from collections import defaultdict
+        self._extra = defaultdict(list)
+        for point in self.points:
+            for key, value in point.extra.items():
+                self._extra[key].append(value)
+        self._extra = dict(self._extra)
 
-        g = dgl.batch(gs)
-        y = torch.tensor(ys, dtype=torch.float32)[:, None]
-        return g, y
+    @property
+    def extra(self):
+        if self._extra is None:
+            self._construct_extra()
+        return self._extra
 
-    @staticmethod
-    def batch_of_g(points):
-        # initialize results
-        gs = []
+    def batch(self, points=None, by=['g', 'y']):
```

Comment:
Tested it. It does work. Sure, I can make it static. I think at some point I was using `self`, but no longer.

Reply:
oh, it does use `self`. if no molecules are provided, it uses self to populate molecules. should I remove that bit?

Reply:
yeah let's use `@staticmethod` and no self

Reply:
by definition, `collate_fn` that you feed into a `dataloader` should be static

Reply:
okay. so then we _always_ require that molecules are provided to the `collate_fn`.

Reply:
I made the `batch` function as we know it a private function for use in the dataloader, and made `batch` that is a class function, so that we can still call `data.batch(by=['g', 'y'])`